### ISSUE TITLE
Skip IsSafeToDeferFormatting when possible

### ISSUE THREAD

**Issue by 304NotModified** [NLog/NLog#1987]
Performance improvement,

skip `IsSafeToDeferFormatting` when by passing enough context. See https://github.com/NLog/NLog/issues/1594

**Comment by snakefoot** [comment]
Currently we have methods like this one:

`public void Trace<TArgument>(string message, TArgument argument)`

One could add an extra version with:

```
[CLSCompliant(false)]
public void Trace<TArgument>(string message, TArgument argument) where TArgument : IConvertible
```

This extra version guarantees one doesn't have to check `IsSafeToDeferFormatting`

**Comment by 304NotModified** [comment]
could we make an overload with only a different `where`?

**Comment by snakefoot** [comment]
> could we make an overload with only a different `where` ?

Don't understand your suggestion. IConvertible is special because it covers the native immutable types (Primitives + String).

**Comment by 304NotModified** [comment]
```
public void Trace<TArgument>(string message, TArgument argument)
public void Trace<TArgument>(string message, TArgument argument) where TArgument : IConvertible
```

both have the same signature  (not sure if the `where` is part of the signature)

**Comment by 304NotModified** [comment]
> To be more specific: in C# the signature consists of the methods:

> - name
> - number of type parameters
> - number of formal parameters
> - type of each formal parameter
> - out/ref/value-ness of each formal parameter
> - with the following additional notes:

> - generic type parameter constraints are not part of the signature

https://stackoverflow.com/questions/8808703/method-signature-in-c-sharp

**Comment by snakefoot** [comment]
You are right:  [Constraints are not part of the signature](http://blogs.msdn.com/ericlippert/archive/2009/12/10/constraints-are-not-part-of-the-signature.aspx)

Well just a random idea.

**Comment by 304NotModified** [comment]
maybe we could do it with this trick (never tested)

```
public void Trace<TArgument>(string message, TArgument argument, bool fast = false) 
public void Trace<TArgument>(string message, TArgument argument) where TArgument : IConvertible
```

(breaking change)

as optional parameters are added add compile stage
